### PR TITLE
Handover SAP elder candidates checks

### PR DESCRIPTION
**Pull request by grumbach**
- Check candidate elders in SAPs from received votes, by correlating them with our membership knowledge
- Add membership generation in SAPs
- Remove membership generation from DKG `Session`, move it to the `DkgSessionId`
- Move all SapCandidates checks to the same place in [here](https://github.com/maidsafe/safe_network/pull/1215/files/c95030e63ae554b2169fd877054b1ab30558b008#diff-1a1f60894e2571c5b9878e6dd6a3400756d27de54479d87a44d7a82c475f381f)

### CODE REVIEW

File: sn_node/src/node/core/messaging/handling/dkg.rs [hunk 1]
```diff
@@ -56,12 +56,7 @@ impl Node {
             peers.push(peer);
         }
 
-        let membership_generation = if let Some(membership) = &*self.membership.read().await {
-            membership.generation()
-        } else {
-            // 0 is the genesis. Which avoids DKG entirely
-            1_u64
-        };
+        let membership_generation = session_id.membership_gen;
```
Comment (grumbach) [inline sn_node/src/node/core/messaging/handling/dkg.rs:60]: There was a race condition here when the generation changed between here and when the DKG session ID is created.

File: sn_node/src/node/handover/handover_consensus.rs [hunk 2]
```diff
@@ -175,58 +167,6 @@ impl Handover {
...
-        // check that the candidates are the oldest in their membership gen
-        // NB TODO check that the sap is valid (either latest candidates or in recent history)
-        if true {
-            Ok(())
-        } else {
-            Err(Error::InvalidSapCandidates)
-        }
-    }
-
-    pub(crate) fn validate_proposal(&self, proposal: SapCandidate) -> Result<()> {
```
Comment (grumbach) [inline sn_node/src/node/handover/handover_consensus.rs:199]: These prefix checks where moved together with the SAP checking code

File: sn_interface/src/network_knowledge/mod.rs [hunk 3]
```diff
@@ -92,7 +92,7 @@ pub const fn supermajority(group_size: usize) -> usize {
     1 + group_size * 2 / 3
 }
 
-pub fn split(
+pub fn partition_by_prefix(
```
Comment (davidrusu) [inline sn_interface/src/network_knowledge/mod.rs:95]: :+1:

**Review by davidrusu — APPROVED** [review]
:+1: looks great

**Pull request closed without merging**